From 410ab96388765422af2b11a76d1a6f7f958bbb99 Mon Sep 17 00:00:00 2001
From: Jianqun Xu <jay.xu@rock-chips.com>
Date: Fri, 8 Oct 2021 16:54:22 +0800
Subject: [PATCH 21/30] pinctrl: rockchip: add DRV_TYPE_IO_SMIC

Add DRV_TYPE_IO_SMIC as a drive strength type for soc produced by smic.

Fixes: d436fb0b1118 ("pinctrl: rockchip: Add rk3308b-s support")

Signed-off-by: Jianqun Xu <jay.xu@rock-chips.com>
Change-Id: I7d2830899a371a7fcf2324303bebbc76a337831e
---
 drivers/pinctrl/pinctrl-rockchip.c | 46 +++++++++++++++++---------------------
 1 file changed, 21 insertions(+), 25 deletions(-)

diff --git a/drivers/pinctrl/pinctrl-rockchip.c b/drivers/pinctrl/pinctrl-rockchip.c
index fe5dc0c..41e2aa7 100644
--- a/drivers/pinctrl/pinctrl-rockchip.c
+++ b/drivers/pinctrl/pinctrl-rockchip.c
@@ -107,6 +107,7 @@ enum rockchip_pin_drv_type {
 	DRV_TYPE_IO_3V3_ONLY,
 	DRV_TYPE_IO_WIDE_LEVEL,
 	DRV_TYPE_IO_NARROW_LEVEL,
+	DRV_TYPE_IO_SMIC,
 	DRV_TYPE_MAX
 };
 
@@ -2266,7 +2267,7 @@ static enum rockchip_pin_drv_type rk3308_calc_drv_reg_and_bit(
 	*bit = (pin_num % RK3288_DRV_PINS_PER_REG);
 	*bit *= RK3288_DRV_BITS_PER_PIN;
 
-	return DRV_TYPE_IO_DEFAULT;
+	return soc_is_rk3308bs() ? DRV_TYPE_IO_SMIC : DRV_TYPE_IO_DEFAULT;
 }
 
 #define RK3366_PULL_GRF_OFFSET		0x110
@@ -2568,7 +2569,8 @@ static int rockchip_perpin_drv_list[DRV_TYPE_MAX][8] = {
 	{ 4, 6, 8, 10, 12, 14, 16, 18 },
 	{ 4, 7, 10, 13, 16, 19, 22, 26 },
 	{ 0, 6, 12, 18, -1, -1, -1, -1 },
-	{ 4, 8, 12, 16, -1, -1, -1, -1 }
+	{ 4, 8, 12, 16, -1, -1, -1, -1 },
+	{ 0, 2, 4, 6, 6, 8, 10, 12 }
 };
 
 static int rockchip_get_drive_perpin(struct rockchip_pin_bank *bank,
@@ -2714,7 +2716,7 @@ static int rockchip_set_drive_perpin(struct rockchip_pin_bank *bank,
 	struct rockchip_pin_ctrl *ctrl = info->ctrl;
 	struct regmap *regmap, *extra_regmap;
 	int reg, ret, i;
-	u32 data, temp, rmask, rmask_bits;
+	u32 data, temp, rmask, rmask_bits = 0;
 	u8 bit, extra_bit;
 	int drv_type, extra_drv_type = 0;
 	int extra_value, extra_reg;
@@ -2751,32 +2753,26 @@ static int rockchip_set_drive_perpin(struct rockchip_pin_bank *bank,
 		return ret;
 	}
 
-	if (soc_is_rk3308bs()) {
-		if (ret == 0)
-			ret = 1;
-		else if (ret == 1)
-			ret = 2;
-		else if (ret == 2)
-			ret = 5;
-		else if (ret == 3)
-			ret = 6;
-
-		data = 0x3 << (bit + 16);
-		rmask = data | (data >> 16);
-		data |= ((ret & 0x3) << bit);
+	switch (drv_type) {
+	case DRV_TYPE_IO_SMIC:
+		if (ctrl->type == RK3308) { /* RK3308B-S */
+			int regval = ret;
 
-		ret = regmap_update_bits(regmap, reg, rmask, data);
-		if (ret)
-			return ret;
+			data = 0x3 << (bit + 16);
+			data |= ((regval & 0x3) << bit);
 
-		rk3308_calc_slew_rate_reg_and_bit(bank, pin_num, &regmap, &reg, &bit);
-		data = BIT(bit + 16) | (((ret > 3) ? 1 : 0) << bit);
-		rmask = BIT(bit + 16) | BIT(bit);
+			ret = regmap_write(regmap, reg, data);
+			if (ret < 0)
+				return ret;
 
-		return regmap_update_bits(regmap, reg, rmask, data);
-	}
+			rk3308_calc_slew_rate_reg_and_bit(bank, pin_num, &regmap, &reg, &bit);
+			data = BIT(bit + 16) | (((regval > 3) ? 1 : 0) << bit);
 
-	switch (drv_type) {
+			return regmap_write(regmap, reg, data);
+		}
+
+		dev_err(info->dev, "unsupported type DRV_TYPE_IO_SMIC\n");
+		return -EINVAL;
 	case DRV_TYPE_IO_1V8_3V0_AUTO:
 	case DRV_TYPE_IO_3V3_ONLY:
 		rmask_bits = RK3399_DRV_3BITS_PER_PIN;
-- 
2.7.4

